### ISSUE TITLE
greenlet kill before it starts running is ignored

### ISSUE THREAD

**Issue by jikamens** [gevent/gevent#330]
This program:

    import gevent

    def messenger1():
        gevent.sleep(0)
        print 'Hello, world.'

    def messenger2():
        print 'Hello again, world.'

    gr1 = gevent.spawn(messenger1)
    gr2 = gevent.spawn(messenger2)
    gevent.kill(gr1)
    gevent.kill(gr2)
    gevent.joinall([gr1, gr2])

Should print nothing, since both greenlets are killed before they start. Unfortunately, on Linux, it prints this:

    Hello again, world.

In other words, if you kill a greenlet before the first time something switches to it, the kill is delayed until gevent switches to another greenlet and then back to it.

This behavior does not strike me as intuitively correct. I think if the greenlet hasn't started running yet and the program kills it, it should just go away.

**Comment by jikamens** [comment]
Thanks for the fix!

**Comment by sublee** [comment]
``successful()`` of killed greenlet is ``True``.  I think its behavior has been changed from this issue fix.  Is it intended behavior?

```python
>>> g = gevent.spawn(gevent.sleep, 10)
>>> g.join(0)  # started
>>> g.kill()
>>> g.successful()
True
>>> g.value
GreenletExit()
>>> g.exception
```

```python
>>> g = gevent.spawn(gevent.sleep, 10)
>>> g.kill()
>>> g.successful()
True
>>> g.value
GreenletExit()
>>> g.exception
```

**Comment by jamadden** [comment]
I don't see a behaviour change from 1.0.2:

```
$ python
Python 2.7.10 (default, May 25 2015, 13:06:17)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import gevent
>>> print(gevent.__version__)
1.0.2
>>>
>>> def run():
...     gevent.sleep(10)
...
>>> g = gevent.spawn(run)
>>> g.join(0)
>>> g.kill()
>>> g.successful()
True
>>> g.exception
>>> g.value
GreenletExit()
>>>
>>> g = gevent.spawn(run)
>>> g.kill()
>>> g.successful()
True
>>> g.exception
>>> g.value
GreenletExit()
```

**Comment by sublee** [comment]
I'm sorry, I was confused my problem.  The problem occurs when killing a greenlet with an exception instead of ``GreenletExit``.

At gevent 1.0.2 or 1.1a1:

```python
>>> g = gevent.spawn(f)
>>> g.kill(ZeroDivisionError)
Traceback (most recent call last):
  ...
ZeroDivisionError
>>> g.successful()
False
```

But at 1.1a2:

```python
>>> g = gevent.spawn(f)
>>> g.kill(ZeroDivisionError)
>>> g.successful()
True
>>> g.get()  # returns GreenletExit instead of raising ZeroDivisionError.
GreenletExit()
>>> g.value
GreenletExit()
>>> g.exception is None
True
```

**Comment by jamadden** [comment]
Ok, thank you. That definitely does look like an unexpected albeit
explainable change in behavior. I'll look into it when I can.


On Jul 9, 2015, at 20:46, Heungsub Lee <notifications@github.com> wrote:

I'm sorry, I was confused my problem. The problem occurs when killing a
greenlet with an exception instead of GreenletExit.

At gevent 1.0.2 or 1.1a1:

>>> g = gevent.spawn(f)>>> g.kill(ZeroDivisionError)
Traceback (most recent call last):
  ...ZeroDivisionError>>> g.successful()False

But at 1.1a2:

>>> g = gevent.spawn(f)>>> g.kill(ZeroDivisionError)>>> g.successful()True>>> g.get()  # returns GreenletExit instead of raising ZeroDivisionError.
GreenletExit()>>> g.value
GreenletExit()>>> g.exception is NoneTrue

—
Reply to this email directly or view it on GitHub
<https://github.com/gevent/gevent/issues/330#issuecomment-120197860>.
